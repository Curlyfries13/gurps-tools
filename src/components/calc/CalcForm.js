import React from 'react';

import TextInput from '../common/TextInput';
import ArmorForm from './ArmorForm';

const CalcForm = ({}) => {
  return (
    <div className='container card'>
      <div className='row g-2 m-2 justify-content-center'>
        <div className='col-md-2 col-6'>
          <TextInput name='HP' id='HP' label='HP' accLabel='Hit Points' />
        </div>
        <div className='col-md-2 col-6 '>
          <TextInput
            name='DR'
            id='DR-sum'
            label='DR'
            accLabel='Damage Resistence'
          />
        </div>
      </div>
      <ArmorForm name='test-armor' id='test-armor-1' />
    </div>
  );
};

export default CalcForm;
