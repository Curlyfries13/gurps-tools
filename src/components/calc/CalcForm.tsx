import React, { useState, useEffect } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Subject } from 'rxjs';

import { addArmor, removeArmor } from '/src/redux/actions/armorActions';
import { resetHP } from '/src/redux/actions/characterActions';
import { RootState } from '/src/redux';

import { Armor } from '/src/types';
import TextInput from '../common/TextInput';
import ArmorForm from './ArmorForm';
import SummaryDR from './SummaryDR';
import DamageTray from './DamageTray';
import HPField from './HPField';

const CalcForm = ({
  armorStack,
  hp,
  currHP,
  summaryDR,
  addArmor,
  removeArmor,
  resetHP,
}: Props) => {
  // This is a crude way to generate unique id's. Use the largest known ID as
  // the starting point.
  const [idCount, setIdCount] = useState(
    armorStack.length === 0
      ? 0
      : armorStack.reduce((prev: Armor, current: Armor) => {
          return prev.id > current.id ? prev : current;
        }).id
  );

  useEffect(() => {
    if (armorStack.length > 0) {
      let value = armorStack.reduce((acc: number, armor: Armor) => {
        return acc + armor.dr;
      }, 0);
      setDR(value);
    }
  }, [armorStack]);

  const collapseSubject = new Subject<void>();

  const handleCollapse = () => {
    collapseSubject.next();
  };

  // only used for summary DR convenience interaction
  const [dr, setDR] = useState(summaryDR ? summaryDR : 0);

  const defaultArmor: Armor = {
    id: idCount,
    order: 0,
    name: 'Armor',
    dr: 0,
    ablative: false,
    ablateBase: 0,
    vulnerabilities: [],
  };

  function handleAddArmor() {
    if (armorStack.length === 0) {
      // Set DR to current DR if initializing stack
      addArmor({ ...defaultArmor, dr, order: armorStack.length });
    } else {
      addArmor({ ...defaultArmor, order: armorStack.length });
    }
    setIdCount(idCount + 1);
  }

  function handleReset() {
    resetHP();
  }

  return (
    <div className='container card'>
      <div className='row g-2 m-2'>
        <div className='col-md-4 col-6'>
          <TextInput id='Name' label='Name' accLabel='Name' />
        </div>
      </div>
      <div className='row g-2 m-2 justify-content-center'>
        <div className='col-md-4 col-6'>
          <HPField />
        </div>
        <div className='col-md-4 col-6'>
          <SummaryDR propDR={dr} setPropDR={setDR} />
        </div>
      </div>
      <div className='row g-2 mb-2'>
        <div className='col-md-1 col-2 mx-1'>
          <button className='btn btn-outline-success' onClick={handleAddArmor}>
            <i className='fs-3 d-flex align-items-top bi bi-shield-fill-plus'></i>
          </button>
        </div>
        <div className='col mx-1'>
          <button
            className='btn btn-outline-primary mx-1'
            onClick={handleCollapse}
            disabled={armorStack.length === 0}
          >
            Collapse All
          </button>
          <button
            className='btn btn-outline-primary mx-1'
            onClick={handleReset}
          >
            Reset
          </button>
        </div>
        <div className='col-md-4 col-4 mx-1'></div>
      </div>
      {armorStack
        .sort((a: Armor, b: Armor) => a.order - b.order)
        .map((armor: Armor) => {
          return (
            <ArmorForm
              key={armor.id}
              armor={armor}
              removeArmor={removeArmor}
              collapse={collapseSubject}
            ></ArmorForm>
          );
        })}
      <div className='card-body'>
        <DamageTray />
      </div>
    </div>
  );
};

interface OwnProps {
  summaryDR: number;
}

function mapStateToProps(state: RootState, ownProps: OwnProps) {
  return {
    armorStack: state.armorStack,
    summaryDR: ownProps.summaryDR,
  };
}

const mapDispatchToProps = {
  addArmor,
  removeArmor,
  resetHP,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type Props = ConnectedProps<typeof connector>;

export default connector(CalcForm);
