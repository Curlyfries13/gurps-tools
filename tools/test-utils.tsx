// Using code from https://redux.js.org/recipes/writing-tests#connected-components
import React from 'react';

// test-utils.js
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer
//
// console.log(React.createElement);

import * as reducer from 'src/redux/reducers';


const customRender = (
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper =({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
