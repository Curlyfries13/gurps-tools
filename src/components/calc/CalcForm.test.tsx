import React from 'react';

import { RenderResult } from '@testing-library/react';
import CalcForm from './CalcForm';

import { render, fireEvent, screen } from 'tools/test-utils';
import store from 'src/redux';

const mockState = {
  armorStack: [],
  displayHP: '0',
  hp: 0,
  displayCurrHP: '0',
  currHp: 0,
  diceMode: false,
  damageExpression: '0',
  expressionIsValid: true,
  log: [],
  currentLogEntry: '',
};

it('matches snapshot', async () => {
  const component: RenderResult = render(<CalcForm />, {
    initialState: mockState,
  });
  expect(component).toMatchSnapshot();
});

it('adds armor components', () => {
  const component: RenderResult = render(<CalcForm />, {
    initialState: mockState,
  });
  fireEvent.click(screen.getByRole('button', { name: /add armor/i }));
  fireEvent.click(screen.getByRole('button', { name: /add armor/i }));

  const test = screen.getAllByRole('listitem', { name: /armor section/i });
  expect(
    screen.getAllByRole('listitem', { name: /armor section/i }).length
  ).toBe(2);
});

// Integration test between CalcForm and Armor sub-components
it('collapses armor components', () => {
  const component: RenderResult = render(<CalcForm />, {
    initialState: mockState,
  });
  // TODO: fix state carryover
  const test = screen.queryAllByRole('listitem', { name: /armor section/i });
  fireEvent.click(screen.getByRole('button', { name: /add armor/i }));

  fireEvent.click(screen.getByRole('button', { name: /collapse all/i }));
  expect(
    screen.queryAllByRole('group', { name: /collapse group/i }).length
  ).toBe(0);
});
