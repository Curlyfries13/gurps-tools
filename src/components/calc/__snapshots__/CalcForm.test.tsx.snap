// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ cQyOIl card"
      >
        <div
          class="container"
        >
          <div
            class="row g-2 m-2"
          >
            <div
              class="col-md-4 col-6"
            >
              <label
                class="form-label"
                for="Name"
              >
                Name
              </label>
              <input
                aria-label="Name"
                class="form-control"
                id="Name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="row g-2 m-2 justify-content-center"
          >
            <div
              class="col-md-4 col-6"
            >
              <label
                class="col col-form-label d-flex justify-content-center"
                for="max-hp"
              >
                HP
              </label>
              <div
                class="d-flex"
              >
                <div>
                  <input
                    aria-label="current hit points"
                    class="form-control"
                    id="current-hp"
                    type="text"
                    value="0"
                  />
                </div>
                <p
                  class="fs-4 mx-2 text-muted"
                >
                  /
                </p>
                <div>
                  <input
                    aria-label="total hit points"
                    class="form-control"
                    id="max-hp"
                    type="text"
                    value="0"
                  />
                </div>
              </div>
            </div>
            <div
              class="col-md-4 col-6"
            >
              <label
                class="form-label d-flex justify-content-center"
                for="DR-sum"
              >
                DR
              </label>
              <input
                aria-label="Summary Damage Resistence"
                class="form-control"
                id="DR-sum"
                type="text"
                value="0"
              />
            </div>
          </div>
          <div
            class="row g-2 mb-2"
          >
            <div
              class="col-md-1 col-2 mx-1"
            >
              <button
                aria-label="add armor"
                class="btn btn-outline-success"
              >
                <i
                  class="fs-3 d-flex align-items-top bi bi-shield-fill-plus"
                />
              </button>
            </div>
            <div
              class="col mx-1"
            >
              <button
                class="btn btn-outline-primary mx-1"
                disabled=""
              >
                Collapse All
              </button>
              <button
                class="btn btn-outline-primary mx-1"
              >
                Reset
              </button>
            </div>
            <div
              class="col-md-4 col-4 mx-1"
            />
          </div>
          <div
            aria-label="armor stack"
            role="list"
          />
          <div
            class="card-body"
          >
            <div
              class="row m-2"
            >
              <div
                class="col col-6 offset-3 d-grid"
              >
                <button
                  class="btn btn-danger"
                >
                  <label>
                    Damage!
                  </label>
                  <i
                    class="bi bi-exclamation-triangle-fill"
                  />
                </button>
              </div>
            </div>
            <div
              class="row m-2"
            >
              <div
                class="col col-6 offset-3 d-grid"
              >
                <label
                  class="col col-form-label d-flex justify-content-center"
                  for="dice-trayvalue"
                >
                  Damage Value
                </label>
                <input
                  aria-label="value for damage calculation"
                  class="form-control"
                  id="dice-trayvalue"
                  type="tet"
                  value="0"
                />
                <div
                  class="row m-2"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ cQyOIl card"
    >
      <div
        class="container"
      >
        <div
          class="row g-2 m-2"
        >
          <div
            class="col-md-4 col-6"
          >
            <label
              class="form-label"
              for="Name"
            >
              Name
            </label>
            <input
              aria-label="Name"
              class="form-control"
              id="Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="row g-2 m-2 justify-content-center"
        >
          <div
            class="col-md-4 col-6"
          >
            <label
              class="col col-form-label d-flex justify-content-center"
              for="max-hp"
            >
              HP
            </label>
            <div
              class="d-flex"
            >
              <div>
                <input
                  aria-label="current hit points"
                  class="form-control"
                  id="current-hp"
                  type="text"
                  value="0"
                />
              </div>
              <p
                class="fs-4 mx-2 text-muted"
              >
                /
              </p>
              <div>
                <input
                  aria-label="total hit points"
                  class="form-control"
                  id="max-hp"
                  type="text"
                  value="0"
                />
              </div>
            </div>
          </div>
          <div
            class="col-md-4 col-6"
          >
            <label
              class="form-label d-flex justify-content-center"
              for="DR-sum"
            >
              DR
            </label>
            <input
              aria-label="Summary Damage Resistence"
              class="form-control"
              id="DR-sum"
              type="text"
              value="0"
            />
          </div>
        </div>
        <div
          class="row g-2 mb-2"
        >
          <div
            class="col-md-1 col-2 mx-1"
          >
            <button
              aria-label="add armor"
              class="btn btn-outline-success"
            >
              <i
                class="fs-3 d-flex align-items-top bi bi-shield-fill-plus"
              />
            </button>
          </div>
          <div
            class="col mx-1"
          >
            <button
              class="btn btn-outline-primary mx-1"
              disabled=""
            >
              Collapse All
            </button>
            <button
              class="btn btn-outline-primary mx-1"
            >
              Reset
            </button>
          </div>
          <div
            class="col-md-4 col-4 mx-1"
          />
        </div>
        <div
          aria-label="armor stack"
          role="list"
        />
        <div
          class="card-body"
        >
          <div
            class="row m-2"
          >
            <div
              class="col col-6 offset-3 d-grid"
            >
              <button
                class="btn btn-danger"
              >
                <label>
                  Damage!
                </label>
                <i
                  class="bi bi-exclamation-triangle-fill"
                />
              </button>
            </div>
          </div>
          <div
            class="row m-2"
          >
            <div
              class="col col-6 offset-3 d-grid"
            >
              <label
                class="col col-form-label d-flex justify-content-center"
                for="dice-trayvalue"
              >
                Damage Value
              </label>
              <input
                aria-label="value for damage calculation"
                class="form-control"
                id="dice-trayvalue"
                type="tet"
                value="0"
              />
              <div
                class="row m-2"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
